designdocs/user-handles.md

```diff
@@ -0,0 +1,326 @@
+```
+
+### Retire user handle
+
+As a network frequency should allow users to retire their handles. This extrinsic will allow users to retire their handles. Retired handles will be available for reuse after a time period set by governance.
+
+``` rust
+Input
+
+* origin - must be a signed origin
+* owner_key - the public key of the user/provider. This is used to verify the signature and resolve the MSA ID.
+* proof_of_ownership - the proof of ownership of the handle to be retired, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+
+* Event - `MsaHandleRetired` with the handle
+
+* Errors - 
+    * `HandleDoesNotExist` if the handle does not exist
+    * `Unauthorized` if the signature is invalid
+
+Validation requirements
+* MSA ID must exist. Signing keys must resolve to MSA ID.
+* handle must exist and be owned by the MSA ID.
+
+Signature requirements
+
+The extrinsic must be signed by the user/provider private key. The signature must be verified on-chain to ensure that the user is the owner of the private key.
+```
+
+**Note:**
+
+* If retiring a ```Pays::No transaction```? then, we could skip both owner_key and proof_of_ownership.
+* If retiring an MSA (```retire_msa``` call), frequency should also retire the handle associated with the MSA.
+
+### Change handle
+
+As a network frequency should allow users to change their handles. This extrinsic will allow users to change their handles. Retired handles will be available for reuse after a time period set by governance.
+
+``` rust
+Input
+* origin - must be a signed origin
+* owner_key - the public key of the user/provider. This is used to verify the signature and resolve the MSA ID.
+* proof_of_new_ownership - the proof of ownership of the new handle, ```MsaHandlePayload``` signed by the user/provider private key
+
+Output
+* Event - `MsaHandleChanged` with the old handle and the new handle
+* Errors - 
+    * `HandleDoesNotExist` if the handle does not exist
+    * `Unauthorized` if the signature(s) is invalid
+    * `InvalidMsaHandle` if the handle is invalid
+
+Validation requirements
+* MSA ID must exist. Signing keys must resolve to MSA ID.
+* old handle must exist and be owned by the MSA ID.
+* new handle must follow handle guidelines.
+* new handle must be unique.
+* new handle must be available.
+* new suffix must be available.
+* new suffix must be in the range of the handle.
+
+Signature requirements
+
+The extrinsic must be signed by the user private key. The signature must be verified on-chain to ensure that the user is the owner of the private key. The signature must also include the old handle and the new handle to prevent unauthorized handle changes.
+```
+
+**Note** :
+
+* Change handle should also retire the old handle and suffix.
+* These are possible extrinsics and inputs required for the handle feature. The exact extrinsics and inputs may change based on the implementation.
```

Comment:
What about transferring ownership?

Reply:
Not sure if it got added to the list, but one of the design goals is that we do not have a secondary market for handles. So a explicit transfer is not possible intentionally.

Reply:
Can we prevent this from happening? I can create accounts with handles associated with them and sell the MSA.

Reply:
Name squatting is hard to prevent but since we have suffixes and a window to control what range of suffixes are being allowed by chain so that takes care of things like TM/Squating etc to some degree.

However, now with updated path, we are not moving the burden of choosing suffixes to user but rather going to put any randomly available suffix to desired user handle

Full Handle can be retired and left to be picked when it becomes available again